{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import sys\n",
    "import pandas as pd"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "TEMP_DATA = \"data/HGTDB/temp_data/\"\n",
    "PATH_TO_PREPROCESSED_HGTDB_FILES = 'data/HGTDB/preprocessed_data/'\n",
    "PATH_TO_PREPROCESSED_HGTDB_FILES_2 = 'data/HGTDB/preprocessed_data_2/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sanity check for preprocessed data!\n",
    "for csv_file in os.listdir(PATH_TO_PREPROCESSED_HGTDB_FILES):\n",
    "    df_1 = pd.read_csv(os.path.join(PATH_TO_PREPROCESSED_HGTDB_FILES, csv_file), index_col='ID')\n",
    "    df_2 = pd.read_csv(os.path.join(PATH_TO_PREPROCESSED_HGTDB_FILES_2, csv_file), index_col='ID')\n",
    "    \n",
    "    if len(df_1) != len(df_2):\n",
    "        print(f'len of 1 {len(df_1)} and len of 2 {len(df_2)}')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_extraneity(dataframe_location):\n",
    "    sim1 = dataframe_location['Sim1']\n",
    "    sim2 = dataframe_location['Sim2']\n",
    "    sim3 = dataframe_location['Sim3']\n",
    "    simt = dataframe_location['SimT']\n",
    "    if sim1 * sim3 > 0 and (abs(sim1)>= 1 or abs(sim3)>= 1):\n",
    "        #total_extraneous = total_extraneous + 1\n",
    "        extraneous = True\n",
    "    elif abs(simt) >= 1:\n",
    "        #total_extraneous = total_extraneous + 1\n",
    "        extraneous = True\n",
    "    else:\n",
    "        extraneous = False\n",
    "    return extraneous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drop_df(dataframe):\n",
    "    dataframe.dropna(inplace=True)\n",
    "    if dataframe.isna().sum().sum() >0:\n",
    "        print(df[df.isna().any(axis=1)])\n",
    "    dataframe = dataframe.bfill(axis='columns')\n",
    "    if dataframe.isna().sum().sum() >0:\n",
    "        print(df[df.isna().any(axis=1)])\n",
    "    return dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "okg\n",
      "spneu1.csv\n",
      "Total hgt counted 258\n",
      "Total extraneous counted 329\n",
      "Total overlap 148\n",
      "okg\n",
      "llac.csv\n",
      "Total hgt counted 90\n",
      "Total extraneous counted 333\n",
      "Total overlap 57\n",
      "okg\n",
      "ypestis.csv\n",
      "Total hgt counted 316\n",
      "Total extraneous counted 466\n",
      "Total overlap 189\n",
      "okg\n",
      "fnucl.csv\n",
      "Total hgt counted 40\n",
      "Total extraneous counted 326\n",
      "Total overlap 15\n",
      "okg\n",
      "bbur.csv\n",
      "Total hgt counted 10\n",
      "Total extraneous counted 133\n",
      "Total overlap 6\n",
      "okg\n",
      "sent.csv\n",
      "Total hgt counted 551\n",
      "Total extraneous counted 548\n",
      "Total overlap 316\n",
      "okg\n",
      "sau2.csv\n",
      "Total hgt counted 119\n",
      "Total extraneous counted 415\n",
      "Total overlap 72\n",
      "okg\n",
      "cjen.csv\n",
      "Total hgt counted 78\n",
      "Total extraneous counted 238\n",
      "Total overlap 34\n",
      "okg\n",
      "bsub.csv\n",
      "Total hgt counted 552\n",
      "Total extraneous counted 552\n",
      "Total overlap 354\n",
      "okg\n",
      "atum2c1.csv\n",
      "Total hgt counted 142\n",
      "Total extraneous counted 446\n",
      "Total overlap 98\n",
      "okg\n",
      "ecoli3.csv\n",
      "Total hgt counted 593\n",
      "Total extraneous counted 671\n",
      "Total overlap 346\n",
      "okg\n",
      "vvul2c2.csv\n",
      "Total hgt counted 207\n",
      "Total extraneous counted 279\n",
      "Total overlap 106\n",
      "okg\n",
      "bhal.csv\n",
      "Total hgt counted 304\n",
      "Total extraneous counted 556\n",
      "Total overlap 172\n",
      "okg\n",
      "xcamp.csv\n",
      "Total hgt counted 285\n",
      "Total extraneous counted 438\n",
      "Total overlap 205\n",
      "okg\n",
      "vcolc2.csv\n",
      "Total hgt counted 204\n",
      "Total extraneous counted 175\n",
      "Total overlap 98\n",
      "okg\n",
      "cpneu.csv\n",
      "Total hgt counted 58\n",
      "Total extraneous counted 166\n",
      "Total overlap 28\n",
      "okg\n",
      "xcitri.csv\n",
      "Total hgt counted 284\n",
      "Total extraneous counted 441\n",
      "Total overlap 206\n",
      "okg\n",
      "sau3.csv\n",
      "Total hgt counted 131\n",
      "Total extraneous counted 427\n",
      "Total overlap 70\n",
      "okg\n",
      "vvul1c2.csv\n",
      "Total hgt counted 199\n",
      "Total extraneous counted 218\n",
      "Total overlap 106\n",
      "okg\n",
      "pmul.csv\n",
      "Total hgt counted 117\n",
      "Total extraneous counted 212\n",
      "Total overlap 61\n",
      "okg\n",
      "synecho.csv\n",
      "Total hgt counted 211\n",
      "Total extraneous counted 343\n",
      "Total overlap 141\n",
      "okg\n",
      "rconorii.csv\n",
      "Total hgt counted 54\n",
      "Total extraneous counted 222\n",
      "Total overlap 27\n",
      "okg\n",
      "bmelic1.csv\n",
      "Total hgt counted 118\n",
      "Total extraneous counted 255\n",
      "Total overlap 85\n",
      "okg\n",
      "dra1.csv\n",
      "Total hgt counted 86\n",
      "Total extraneous counted 202\n",
      "Total overlap 57\n",
      "okg\n",
      "mpul.csv\n",
      "Total hgt counted 28\n",
      "Total extraneous counted 99\n",
      "Total overlap 12\n",
      "okg\n",
      "cglu.csv\n",
      "Total hgt counted 207\n",
      "Total extraneous counted 400\n",
      "Total overlap 137\n",
      "okg\n",
      "rsolac2.csv\n",
      "Total hgt counted 172\n",
      "Total extraneous counted 174\n",
      "Total overlap 112\n",
      "okg\n",
      "ctep.csv\n",
      "Total hgt counted 267\n",
      "Total extraneous counted 473\n",
      "Total overlap 178\n",
      "okg\n",
      "nmen1.csv\n",
      "Total hgt counted 221\n",
      "Total extraneous counted 239\n",
      "Total overlap 128\n",
      "okg\n",
      "cperf.csv\n",
      "Total hgt counted 75\n",
      "Total extraneous counted 383\n",
      "Total overlap 41\n",
      "okg\n",
      "mtub2.csv\n",
      "Total hgt counted 197\n",
      "Total extraneous counted 537\n",
      "Total overlap 116\n",
      "okg\n",
      "uure.csv\n",
      "Total hgt counted 12\n",
      "Total extraneous counted 84\n",
      "Total overlap 5\n",
      "okg\n",
      "atum1c1.csv\n",
      "Total hgt counted 194\n",
      "Total extraneous counted 416\n",
      "Total overlap 128\n",
      "okg\n",
      "ecoli2.csv\n",
      "Total hgt counted 625\n",
      "Total extraneous counted 651\n",
      "Total overlap 353\n",
      "okg\n",
      "ctra.csv\n",
      "Total hgt counted 36\n",
      "Total extraneous counted 125\n",
      "Total overlap 23\n",
      "okg\n",
      "mlep.csv\n",
      "Total hgt counted 73\n",
      "Total extraneous counted 207\n",
      "Total overlap 46\n",
      "okg\n",
      "mloti.csv\n",
      "Total hgt counted 604\n",
      "Total extraneous counted 952\n",
      "Total overlap 409\n",
      "okg\n",
      "hpyl99.csv\n",
      "Total hgt counted 68\n",
      "Total extraneous counted 181\n",
      "Total overlap 41\n",
      "okg\n",
      "rsola.csv\n",
      "Total hgt counted 356\n",
      "Total extraneous counted 400\n",
      "Total overlap 202\n",
      "okg\n",
      "xfas.csv\n",
      "Total hgt counted 458\n",
      "Total extraneous counted 529\n",
      "Total overlap 270\n",
      "okg\n",
      "nmen2.csv\n",
      "Total hgt counted 206\n",
      "Total extraneous counted 244\n",
      "Total overlap 126\n",
      "okg\n",
      "atum1c2.csv\n",
      "Total hgt counted 114\n",
      "Total extraneous counted 218\n",
      "Total overlap 87\n",
      "okg\n",
      "vvul1c1.csv\n",
      "Total hgt counted 320\n",
      "Total extraneous counted 413\n",
      "Total overlap 206\n",
      "okg\n",
      "btheta.csv\n",
      "Total hgt counted 753\n",
      "Total extraneous counted 673\n",
      "Total overlap 395\n",
      "okg\n",
      "ccres.csv\n",
      "Total hgt counted 135\n",
      "Total extraneous counted 505\n",
      "Total overlap 107\n",
      "okg\n",
      "tteng.csv\n",
      "Total hgt counted 343\n",
      "Total extraneous counted 369\n",
      "Total overlap 163\n",
      "okg\n",
      "spyo2.csv\n",
      "Total hgt counted 157\n",
      "Total extraneous counted 281\n",
      "Total overlap 100\n",
      "okg\n",
      "styp.csv\n",
      "Total hgt counted 446\n",
      "Total extraneous counted 521\n",
      "Total overlap 260\n",
      "okg\n",
      "paer.csv\n",
      "Total hgt counted 307\n",
      "Total extraneous counted 542\n",
      "Total overlap 186\n",
      "okg\n",
      "linno.csv\n",
      "Total hgt counted 164\n",
      "Total extraneous counted 398\n",
      "Total overlap 110\n",
      "okg\n",
      "nsp.csv\n",
      "Total hgt counted 203\n",
      "Total extraneous counted 741\n",
      "Total overlap 111\n",
      "okg\n",
      "bmelic2.csv\n",
      "Total hgt counted 59\n",
      "Total extraneous counted 161\n",
      "Total overlap 33\n",
      "okg\n",
      "cpneu2.csv\n",
      "Total hgt counted 55\n",
      "Total extraneous counted 209\n",
      "Total overlap 32\n",
      "okg\n",
      "dra2.csv\n",
      "Total hgt counted 23\n",
      "Total extraneous counted 21\n",
      "Total overlap 11\n",
      "okg\n",
      "bsp.csv\n",
      "Total hgt counted 0\n",
      "Total extraneous counted 85\n",
      "Total overlap 0\n",
      "okg\n",
      "baphi.csv\n",
      "Total hgt counted 6\n",
      "Total extraneous counted 84\n",
      "Total overlap 6\n",
      "okg\n",
      "cmur.csv\n",
      "Total hgt counted 12\n",
      "Total extraneous counted 128\n",
      "Total overlap 9\n",
      "okg\n",
      "hinf.csv\n",
      "Total hgt counted 87\n",
      "Total extraneous counted 169\n",
      "Total overlap 59\n",
      "okg\n",
      "mgen.csv\n",
      "Total hgt counted 51\n",
      "Total extraneous counted 67\n",
      "Total overlap 14\n",
      "okg\n",
      "vvul2c1.csv\n",
      "Total hgt counted 375\n",
      "Total extraneous counted 429\n",
      "Total overlap 219\n",
      "okg\n",
      "rpxx.csv\n",
      "Total hgt counted 28\n",
      "Total extraneous counted 123\n",
      "Total overlap 10\n",
      "okg\n",
      "ecoli.csv\n",
      "Total hgt counted 359\n",
      "Total extraneous counted 458\n",
      "Total overlap 219\n",
      "okg\n",
      "vcolc1.csv\n",
      "Total hgt counted 234\n",
      "Total extraneous counted 340\n",
      "Total overlap 128\n",
      "okg\n",
      "mtub.csv\n",
      "Total hgt counted 176\n",
      "Total extraneous counted 438\n",
      "Total overlap 97\n",
      "okg\n",
      "spyo.csv\n",
      "Total hgt counted 136\n",
      "Total extraneous counted 231\n",
      "Total overlap 80\n",
      "okg\n",
      "lmono.csv\n",
      "Total hgt counted 184\n",
      "Total extraneous counted 384\n",
      "Total overlap 126\n",
      "okg\n",
      "smel.csv\n",
      "Total hgt counted 179\n",
      "Total extraneous counted 464\n",
      "Total overlap 139\n",
      "okg\n",
      "scoel.csv\n",
      "Total hgt counted 128\n",
      "Total extraneous counted 331\n",
      "Total overlap 60\n",
      "okg\n",
      "tmar.csv\n",
      "Total hgt counted 194\n",
      "Total extraneous counted 257\n",
      "Total overlap 106\n",
      "okg\n",
      "spneu2.csv\n",
      "Total hgt counted 249\n",
      "Total extraneous counted 277\n",
      "Total overlap 156\n",
      "okg\n",
      "aquae.csv\n",
      "Total hgt counted 70\n",
      "Total extraneous counted 164\n",
      "Total overlap 43\n",
      "okg\n",
      "cpneu3.csv\n",
      "Total hgt counted 49\n",
      "Total extraneous counted 166\n",
      "Total overlap 22\n",
      "okg\n",
      "tpal.csv\n",
      "Total hgt counted 78\n",
      "Total extraneous counted 143\n",
      "Total overlap 39\n",
      "okg\n",
      "mpneu.csv\n",
      "Total hgt counted 39\n",
      "Total extraneous counted 82\n",
      "Total overlap 26\n",
      "okg\n",
      "hpyl.csv\n",
      "Total hgt counted 87\n",
      "Total extraneous counted 240\n",
      "Total overlap 60\n",
      "okg\n",
      "caceto.csv\n",
      "Total hgt counted 146\n",
      "Total extraneous counted 530\n",
      "Total overlap 103\n",
      "okg\n",
      "atum2c2.csv\n",
      "Total hgt counted 114\n",
      "Total extraneous counted 233\n",
      "Total overlap 82\n",
      "okg\n",
      "sau1.csv\n",
      "Total hgt counted 105\n",
      "Total extraneous counted 411\n",
      "Total overlap 68\n"
     ]
    }
   ],
   "source": [
    "to_csv = []\n",
    "for csv_file in os.listdir(PATH_TO_PREPROCESSED_HGTDB_FILES):\n",
    "    #df = pd.read_csv(os.path.join(PATH_TO_PREPROCESSED_HGTDB_FILES, csv_file), index_col='ID')\n",
    "    df = pd.read_csv(os.path.join(PATH_TO_PREPROCESSED_HGTDB_FILES_2, csv_file), index_col='ID')\n",
    "    #df_temp_data = pd.read_csv(os.path.join(TEMP_DATA, csv_file.replace('.csv','.tsv')), sep='\\t', index_col=\"Synonym\")\n",
    "    \n",
    "    df = df.drop(columns=[\"FunctionCode\",\"Strand\",\"AADev\",\"Length\",\"GC1\",\"GC2\",\"GC3\",\"GCT\",\"SD1\",\"SD2\",\"SD3\",\"SDT\",\"Mah\"])\n",
    "    \n",
    "    df = drop_df(df)\n",
    "    df['extraneous'] = [check_extraneity(df.iloc[i]) for i in range(len(df))]\n",
    "    #HGT check\n",
    "    hgt_check = df['HGT'] == 1\n",
    "    if (hgt_check[hgt_check==False].count() + hgt_check[hgt_check==True].count()) == len(df):\n",
    "        print('okg')\n",
    "    \n",
    "    \n",
    "    \n",
    "    print(f'{csv_file}')\n",
    "    #print(f'lenght before {len(df_temp_data)} lenght after {len(df_temp_data)}')\n",
    "    print(f\"Total hgt counted { df[df['HGT']==1]['HGT'].count()}\")\n",
    "    print(f\"Total extraneous counted {df[df['extraneous'] == True]['extraneous'].count()}\")\n",
    "    print(f\"Total overlap {df[(df['HGT']==1) & (df['extraneous'] == True)]['HGT'].count()}\")\n",
    "    \n",
    "    #['genome','len','HGT_count','extraneous_count', 'overlap_count']\n",
    "    to_csv.append((csv_file,len(df), df[df['HGT']==1]['HGT'].count(),df[df['extraneous'] == True]['extraneous'].count(), df[(df['HGT']==1) & (df['extraneous'] == True)]['HGT'].count() ))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('spneu1.csv', 2094, 258, 329, 148),\n",
       " ('llac.csv', 2267, 90, 333, 57),\n",
       " ('ypestis.csv', 3885, 316, 466, 189),\n",
       " ('fnucl.csv', 2067, 40, 326, 15),\n",
       " ('bbur.csv', 851, 10, 133, 6),\n",
       " ('sent.csv', 4395, 551, 548, 316),\n",
       " ('sau2.csv', 2714, 119, 415, 72),\n",
       " ('cjen.csv', 1634, 78, 238, 34),\n",
       " ('bsub.csv', 4111, 552, 552, 354),\n",
       " ('atum2c1.csv', 2785, 142, 446, 98),\n",
       " ('ecoli3.csv', 5324, 593, 671, 346),\n",
       " ('vvul2c2.csv', 1697, 207, 279, 106),\n",
       " ('bhal.csv', 4066, 304, 556, 172),\n",
       " ('xcamp.csv', 4181, 285, 438, 205),\n",
       " ('vcolc2.csv', 1093, 204, 175, 98),\n",
       " ('cpneu.csv', 1054, 58, 166, 28),\n",
       " ('xcitri.csv', 4312, 284, 441, 206),\n",
       " ('sau3.csv', 2632, 131, 427, 70),\n",
       " ('vvul1c2.csv', 1565, 199, 218, 106),\n",
       " ('pmul.csv', 2015, 117, 212, 61),\n",
       " ('synecho.csv', 3167, 211, 343, 141),\n",
       " ('rconorii.csv', 1374, 54, 222, 27),\n",
       " ('bmelic1.csv', 2059, 118, 255, 85),\n",
       " ('dra1.csv', 2628, 86, 202, 57),\n",
       " ('mpul.csv', 782, 28, 99, 12),\n",
       " ('cglu.csv', 3040, 207, 400, 137),\n",
       " ('rsolac2.csv', 1676, 172, 174, 112),\n",
       " ('ctep.csv', 2252, 267, 473, 178),\n",
       " ('nmen1.csv', 2079, 221, 239, 128),\n",
       " ('cperf.csv', 2660, 75, 383, 41),\n",
       " ('mtub2.csv', 4187, 197, 537, 116),\n",
       " ('uure.csv', 614, 12, 84, 5),\n",
       " ('atum1c1.csv', 2721, 194, 416, 128),\n",
       " ('ecoli2.csv', 5361, 625, 651, 353),\n",
       " ('ctra.csv', 895, 36, 125, 23),\n",
       " ('mlep.csv', 1605, 73, 207, 46),\n",
       " ('mloti.csv', 6746, 604, 952, 409),\n",
       " ('hpyl99.csv', 1490, 68, 181, 41),\n",
       " ('rsola.csv', 3440, 356, 400, 202),\n",
       " ('xfas.csv', 2766, 458, 529, 270),\n",
       " ('nmen2.csv', 2065, 206, 244, 126),\n",
       " ('atum1c2.csv', 1833, 114, 218, 87),\n",
       " ('vvul1c1.csv', 2972, 320, 413, 206),\n",
       " ('btheta.csv', 4778, 753, 673, 395),\n",
       " ('ccres.csv', 3737, 135, 505, 107),\n",
       " ('tteng.csv', 2588, 343, 369, 163),\n",
       " ('spyo2.csv', 1845, 157, 281, 100),\n",
       " ('styp.csv', 4451, 446, 521, 260),\n",
       " ('paer.csv', 5567, 307, 542, 186),\n",
       " ('linno.csv', 2968, 164, 398, 110),\n",
       " ('nsp.csv', 5366, 203, 741, 111),\n",
       " ('bmelic2.csv', 1139, 59, 161, 33),\n",
       " ('cpneu2.csv', 1112, 55, 209, 32),\n",
       " ('dra2.csv', 368, 23, 21, 11),\n",
       " ('bsp.csv', 564, 0, 85, 0),\n",
       " ('baphi.csv', 544, 6, 84, 6),\n",
       " ('cmur.csv', 909, 12, 128, 9),\n",
       " ('hinf.csv', 1714, 87, 169, 59),\n",
       " ('mgen.csv', 484, 51, 67, 14),\n",
       " ('vvul2c1.csv', 3262, 375, 429, 219),\n",
       " ('rpxx.csv', 835, 28, 123, 10),\n",
       " ('ecoli.csv', 4276, 359, 458, 219),\n",
       " ('vcolc1.csv', 2742, 234, 340, 128),\n",
       " ('mtub.csv', 3927, 176, 438, 97),\n",
       " ('spyo.csv', 1697, 136, 231, 80),\n",
       " ('lmono.csv', 2846, 184, 384, 126),\n",
       " ('smel.csv', 3341, 179, 464, 139),\n",
       " ('scoel.csv', 3030, 128, 331, 60),\n",
       " ('tmar.csv', 1858, 194, 257, 106),\n",
       " ('spneu2.csv', 2043, 249, 277, 156),\n",
       " ('aquae.csv', 1529, 70, 164, 43),\n",
       " ('cpneu3.csv', 1069, 49, 166, 22),\n",
       " ('tpal.csv', 1036, 78, 143, 39),\n",
       " ('mpneu.csv', 689, 39, 82, 26),\n",
       " ('hpyl.csv', 1576, 87, 240, 60),\n",
       " ('caceto.csv', 3672, 146, 530, 103),\n",
       " ('atum2c2.csv', 1876, 114, 233, 82),\n",
       " ('sau1.csv', 2594, 105, 411, 68)]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = 'ignore/data_statistic_report_thesis_HGTDB.csv'\n",
    "\n",
    "\n",
    "with open(file_name,'w') as out:\n",
    "    csv_out=csv.writer(out)\n",
    "    #csv_out.writerow(['genome','type','gene','SD1','SD2','SD3','SDT'])\n",
    "    csv_out.writerow(['genome','len','HGT_count','extraneous_count', 'overlap_count'])\n",
    "    csv_out.writerows(to_csv)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evaluation Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def open_result_json(file_path):\n",
    "    with open(file_path) as payload:\n",
    "        json_payload = json.load(payload)\n",
    "    return json_payload"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 15 ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_genome_csv = 'test_data/15_test_data_RTR.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_list = pd.read_csv(ref_genome_csv)\n",
    "\n",
    "ref_list['arg_folder_target'] = ['results/HGT_DB/analysis/card/'+i +'/' +i +'.json' for i in ref_list['asm_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTDB_taxonomy</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>dl_link</th>\n",
       "      <th>file_path</th>\n",
       "      <th>ori_path</th>\n",
       "      <th>arg_folder_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d__Bacteria;p__Bacillota;c__Bacilli;o__Bacilla...</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>ASM904v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM904v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM904v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM904v1/ASM904v1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d__Bacteria;p__Chlamydiota;c__Chlamydiia;o__Ch...</td>\n",
       "      <td>Chlamydia muridarum str. Nigg</td>\n",
       "      <td>ASM668v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM668v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM668v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM668v1/ASM668v1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d__Bacteria;p__Bacillota;c__Bacilli;o__Lactoba...</td>\n",
       "      <td>Lacticaseibacillus casei DSM 20011 = JCM 1134 ...</td>\n",
       "      <td>ASM82905v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM82905v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM82905v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM82905v1/ASM829...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d__Bacteria;p__Pseudomonadota;c__Gammaproteoba...</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>ASM584v2</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM584v2.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM584v2.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM584v2/ASM584v2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d__Bacteria;p__Bacillota;c__Bacilli;o__Bacilla...</td>\n",
       "      <td>Bacillus anthracis str. 'Ames Ancestor'</td>\n",
       "      <td>ASM844v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM844v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM844v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM844v1/ASM844v1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>d__Bacteria;p__Actinomycetota;c__Actinomycetia...</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>ASM19595v2</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM19595v2.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM19595v2.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM19595v2/ASM195...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>d__Bacteria;p__Bacillota;c__Bacilli;o__Staphyl...</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>ASM1342v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM1342v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM1342v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM1342v1/ASM1342...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>d__Bacteria;p__Pseudomonadota;c__Gammaproteoba...</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>ASM694v2</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM694v2.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM694v2.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM694v2/ASM694v2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>d__Bacteria;p__Pseudomonadota;c__Alphaproteoba...</td>\n",
       "      <td>Rickettsia rickettsii str. 'Sheila Smith'</td>\n",
       "      <td>ASM1822v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM1822v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM1822v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM1822v1/ASM1822...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>d__Bacteria;p__Bacillota;c__Bacilli;o__Lactoba...</td>\n",
       "      <td>Listeria innocua</td>\n",
       "      <td>ASM964857v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/009...</td>\n",
       "      <td>data/prepped/ASM964857v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM964857v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM964857v1/ASM96...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>d__Bacteria;p__Bacillota;c__Bacilli;o__Lactoba...</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "      <td>ASM207683v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002...</td>\n",
       "      <td>data/prepped/ASM207683v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM207683v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM207683v1/ASM20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>d__Bacteria;p__Campylobacterota;c__Campylobact...</td>\n",
       "      <td>Helicobacter pylori</td>\n",
       "      <td>ASM1782153v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/017...</td>\n",
       "      <td>data/prepped/ASM1782153v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM1782153v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM1782153v1/ASM1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>d__Bacteria;p__Aquificota;c__Aquificae;o__Aqui...</td>\n",
       "      <td>Aquifex aeolicus VF5</td>\n",
       "      <td>ASM862v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM862v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM862v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM862v1/ASM862v1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>d__Bacteria;p__Actinomycetota;c__Actinomycetia...</td>\n",
       "      <td>Corynebacterium glutamicum SCgG2</td>\n",
       "      <td>ASM40418v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...</td>\n",
       "      <td>data/prepped/ASM40418v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM40418v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM40418v1/ASM404...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>d__Bacteria;p__Bacillota_A;c__Clostridia;o__Cl...</td>\n",
       "      <td>Clostridium perfringens</td>\n",
       "      <td>ASM2013877v1</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/020...</td>\n",
       "      <td>data/prepped/ASM2013877v1.csv</td>\n",
       "      <td>data/NCBI/sequence_files/ASM2013877v1.fna</td>\n",
       "      <td>results/HGT_DB/analysis/card/ASM2013877v1/ASM2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        GTDB_taxonomy   \n",
       "0   d__Bacteria;p__Bacillota;c__Bacilli;o__Bacilla...  \\\n",
       "1   d__Bacteria;p__Chlamydiota;c__Chlamydiia;o__Ch...   \n",
       "2   d__Bacteria;p__Bacillota;c__Bacilli;o__Lactoba...   \n",
       "3   d__Bacteria;p__Pseudomonadota;c__Gammaproteoba...   \n",
       "4   d__Bacteria;p__Bacillota;c__Bacilli;o__Bacilla...   \n",
       "5   d__Bacteria;p__Actinomycetota;c__Actinomycetia...   \n",
       "6   d__Bacteria;p__Bacillota;c__Bacilli;o__Staphyl...   \n",
       "7   d__Bacteria;p__Pseudomonadota;c__Gammaproteoba...   \n",
       "8   d__Bacteria;p__Pseudomonadota;c__Alphaproteoba...   \n",
       "9   d__Bacteria;p__Bacillota;c__Bacilli;o__Lactoba...   \n",
       "10  d__Bacteria;p__Bacillota;c__Bacilli;o__Lactoba...   \n",
       "11  d__Bacteria;p__Campylobacterota;c__Campylobact...   \n",
       "12  d__Bacteria;p__Aquificota;c__Aquificae;o__Aqui...   \n",
       "13  d__Bacteria;p__Actinomycetota;c__Actinomycetia...   \n",
       "14  d__Bacteria;p__Bacillota_A;c__Clostridia;o__Cl...   \n",
       "\n",
       "                                        organism_name      asm_name   \n",
       "0          Bacillus subtilis subsp. subtilis str. 168      ASM904v1  \\\n",
       "1                       Chlamydia muridarum str. Nigg      ASM668v1   \n",
       "2   Lacticaseibacillus casei DSM 20011 = JCM 1134 ...    ASM82905v1   \n",
       "3           Escherichia coli str. K-12 substr. MG1655      ASM584v2   \n",
       "4             Bacillus anthracis str. 'Ames Ancestor'      ASM844v1   \n",
       "5                    Mycobacterium tuberculosis H37Rv    ASM19595v2   \n",
       "6       Staphylococcus aureus subsp. aureus NCTC 8325     ASM1342v1   \n",
       "7   Salmonella enterica subsp. enterica serovar Ty...      ASM694v2   \n",
       "8           Rickettsia rickettsii str. 'Sheila Smith'     ASM1822v1   \n",
       "9                                    Listeria innocua   ASM964857v1   \n",
       "10                           Streptococcus pneumoniae   ASM207683v1   \n",
       "11                                Helicobacter pylori  ASM1782153v1   \n",
       "12                               Aquifex aeolicus VF5      ASM862v1   \n",
       "13                   Corynebacterium glutamicum SCgG2    ASM40418v1   \n",
       "14                            Clostridium perfringens  ASM2013877v1   \n",
       "\n",
       "                                              dl_link   \n",
       "0   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...  \\\n",
       "1   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "2   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "3   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "4   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "5   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "6   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "7   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "8   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "9   ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/009...   \n",
       "10  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002...   \n",
       "11  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/017...   \n",
       "12  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "13  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000...   \n",
       "14  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/020...   \n",
       "\n",
       "                        file_path                                   ori_path   \n",
       "0       data/prepped/ASM904v1.csv      data/NCBI/sequence_files/ASM904v1.fna  \\\n",
       "1       data/prepped/ASM668v1.csv      data/NCBI/sequence_files/ASM668v1.fna   \n",
       "2     data/prepped/ASM82905v1.csv    data/NCBI/sequence_files/ASM82905v1.fna   \n",
       "3       data/prepped/ASM584v2.csv      data/NCBI/sequence_files/ASM584v2.fna   \n",
       "4       data/prepped/ASM844v1.csv      data/NCBI/sequence_files/ASM844v1.fna   \n",
       "5     data/prepped/ASM19595v2.csv    data/NCBI/sequence_files/ASM19595v2.fna   \n",
       "6      data/prepped/ASM1342v1.csv     data/NCBI/sequence_files/ASM1342v1.fna   \n",
       "7       data/prepped/ASM694v2.csv      data/NCBI/sequence_files/ASM694v2.fna   \n",
       "8      data/prepped/ASM1822v1.csv     data/NCBI/sequence_files/ASM1822v1.fna   \n",
       "9    data/prepped/ASM964857v1.csv   data/NCBI/sequence_files/ASM964857v1.fna   \n",
       "10   data/prepped/ASM207683v1.csv   data/NCBI/sequence_files/ASM207683v1.fna   \n",
       "11  data/prepped/ASM1782153v1.csv  data/NCBI/sequence_files/ASM1782153v1.fna   \n",
       "12      data/prepped/ASM862v1.csv      data/NCBI/sequence_files/ASM862v1.fna   \n",
       "13    data/prepped/ASM40418v1.csv    data/NCBI/sequence_files/ASM40418v1.fna   \n",
       "14  data/prepped/ASM2013877v1.csv  data/NCBI/sequence_files/ASM2013877v1.fna   \n",
       "\n",
       "                                    arg_folder_target  \n",
       "0   results/HGT_DB/analysis/card/ASM904v1/ASM904v1...  \n",
       "1   results/HGT_DB/analysis/card/ASM668v1/ASM668v1...  \n",
       "2   results/HGT_DB/analysis/card/ASM82905v1/ASM829...  \n",
       "3   results/HGT_DB/analysis/card/ASM584v2/ASM584v2...  \n",
       "4   results/HGT_DB/analysis/card/ASM844v1/ASM844v1...  \n",
       "5   results/HGT_DB/analysis/card/ASM19595v2/ASM195...  \n",
       "6   results/HGT_DB/analysis/card/ASM1342v1/ASM1342...  \n",
       "7   results/HGT_DB/analysis/card/ASM694v2/ASM694v2...  \n",
       "8   results/HGT_DB/analysis/card/ASM1822v1/ASM1822...  \n",
       "9   results/HGT_DB/analysis/card/ASM964857v1/ASM96...  \n",
       "10  results/HGT_DB/analysis/card/ASM207683v1/ASM20...  \n",
       "11  results/HGT_DB/analysis/card/ASM1782153v1/ASM1...  \n",
       "12  results/HGT_DB/analysis/card/ASM862v1/ASM862v1...  \n",
       "13  results/HGT_DB/analysis/card/ASM40418v1/ASM404...  \n",
       "14  results/HGT_DB/analysis/card/ASM2013877v1/ASM2...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ASM904v1\n",
      "Total ARG 16\n",
      "Total extraneous counted 544\n",
      "Toal ARG that is extraneous 0\n",
      "ASM668v1\n",
      "Total ARG 1\n",
      "Total extraneous counted 96\n",
      "Toal ARG that is extraneous 0\n",
      "ASM82905v1\n",
      "Total ARG 0\n",
      "Total extraneous counted 316\n",
      "Toal ARG that is extraneous 0\n",
      "ASM584v2\n",
      "Total ARG 53\n",
      "Total extraneous counted 406\n",
      "Toal ARG that is extraneous 7\n",
      "ASM844v1\n",
      "Total ARG 16\n",
      "Total extraneous counted 637\n",
      "Toal ARG that is extraneous 1\n",
      "ASM19595v2\n",
      "Total ARG 9\n",
      "Total extraneous counted 428\n",
      "Toal ARG that is extraneous 1\n",
      "ASM1342v1\n",
      "Total ARG 12\n",
      "Total extraneous counted 342\n",
      "Toal ARG that is extraneous 0\n",
      "ASM694v2\n",
      "Total ARG 32\n",
      "Total extraneous counted 498\n",
      "Toal ARG that is extraneous 0\n",
      "ASM1822v1\n",
      "Total ARG 0\n",
      "Total extraneous counted 172\n",
      "Toal ARG that is extraneous 0\n",
      "ASM964857v1\n",
      "Total ARG 4\n",
      "Total extraneous counted 349\n",
      "Toal ARG that is extraneous 0\n",
      "ASM207683v1\n",
      "Total ARG 7\n",
      "Total extraneous counted 241\n",
      "Toal ARG that is extraneous 1\n",
      "ASM1782153v1\n",
      "Total ARG 4\n",
      "Total extraneous counted 196\n",
      "Toal ARG that is extraneous 0\n",
      "ASM862v1\n",
      "Total ARG 0\n",
      "Total extraneous counted 173\n",
      "Toal ARG that is extraneous 0\n",
      "ASM40418v1\n",
      "Total ARG 2\n",
      "Total extraneous counted 344\n",
      "Toal ARG that is extraneous 0\n",
      "ASM2013877v1\n",
      "Total ARG 6\n",
      "Total extraneous counted 359\n",
      "Toal ARG that is extraneous 0\n"
     ]
    }
   ],
   "source": [
    "to_csv_2 = []\n",
    "for i in range(len(ref_list)):\n",
    "    df = pd.read_csv(ref_list.iloc[i]['file_path'],index_col=0)\n",
    "    arg_genes_json = open_result_json(ref_list.iloc[i]['arg_folder_target'])\n",
    "    \n",
    "    arg_genes = list(arg_genes_json.keys())\n",
    "    \n",
    "    df['extraneous'] = [check_extraneity(df.iloc[i]) for i in range(len(df))]\n",
    "    \n",
    "    print(f\"{ref_list.iloc[i]['asm_name']}\")\n",
    "    print(f\"Total ARG {len(arg_genes)}\")\n",
    "    print(f\"Total extraneous counted {df[df['extraneous'] == True]['extraneous'].count()}\")\n",
    "    arg_genes_df = df.loc[arg_genes]\n",
    "    print(f\"Toal ARG that is extraneous {arg_genes_df[arg_genes_df['extraneous'] == True]['extraneous'].count()}\")\n",
    "    \n",
    "    to_csv_2.append((ref_list.iloc[i]['asm_name'],len(df), len(arg_genes), df[df['extraneous'] == True]['extraneous'].count(), arg_genes_df[arg_genes_df['extraneous'] == True]['extraneous'].count()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('ASM904v1', 4325, 16, 544, 0),\n",
       " ('ASM668v1', 911, 1, 96, 0),\n",
       " ('ASM82905v1', 2777, 0, 316, 0),\n",
       " ('ASM584v2', 4305, 53, 406, 7),\n",
       " ('ASM844v1', 5486, 16, 637, 1),\n",
       " ('ASM19595v2', 4031, 9, 428, 1),\n",
       " ('ASM1342v1', 2892, 12, 342, 0),\n",
       " ('ASM694v2', 4554, 32, 498, 0),\n",
       " ('ASM1822v1', 1345, 0, 172, 0),\n",
       " ('ASM964857v1', 2866, 4, 349, 0),\n",
       " ('ASM207683v1', 2192, 7, 241, 1),\n",
       " ('ASM1782153v1', 1489, 4, 196, 0),\n",
       " ('ASM862v1', 1553, 0, 173, 0),\n",
       " ('ASM40418v1', 3025, 2, 344, 0),\n",
       " ('ASM2013877v1', 2842, 6, 359, 0)]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_csv_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = 'ignore/data_statistic_report_thesis_15.csv'\n",
    "\n",
    "\n",
    "with open(file_name,'w') as out:\n",
    "    csv_out=csv.writer(out)\n",
    "    #csv_out.writerow(['genome','type','gene','SD1','SD2','SD3','SDT'])\n",
    "    csv_out.writerow(['genome','len','ARG','extraneous count', 'overlap'])\n",
    "    csv_out.writerows(to_csv_2)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## outbreak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "outbreak_genome_csv = 'outbreak_genome/outbreak_genome.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "outbreak_list = pd.read_csv(outbreak_genome_csv)\n",
    "\n",
    "# get asm name\n",
    "outbreak_list['asm_name'] = [i.split('_')[3] for i in outbreak_list['file_path']]\n",
    "\n",
    "# get args path\n",
    "roots_arg = ['outbreak_'+i.split('_')[3] for i in outbreak_list['file_path']]\n",
    "outbreak_list['arg_folder_target'] = ['results/HGT_DB/analysis/card/'+i +'/' +i +'.json' for i in roots_arg]\n",
    "\n",
    "\n",
    "# get csv path \n",
    "outbreak_list['file_path_csv'] = [i.replace('.fna','.csv') for i in outbreak_list['file_path']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_path</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>arg_folder_target</th>\n",
       "      <th>file_path_csv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>outbreak_genome/GCA_013459015.1_ASM1345901v1_c...</td>\n",
       "      <td>ASM1345901v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM13459...</td>\n",
       "      <td>outbreak_genome/GCA_013459015.1_ASM1345901v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>outbreak_genome/GCA_013459175.1_ASM1345917v1_c...</td>\n",
       "      <td>ASM1345917v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM13459...</td>\n",
       "      <td>outbreak_genome/GCA_013459175.1_ASM1345917v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>outbreak_genome/GCA_015209985.1_ASM1520998v1_c...</td>\n",
       "      <td>ASM1520998v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM15209...</td>\n",
       "      <td>outbreak_genome/GCA_015209985.1_ASM1520998v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>outbreak_genome/GCA_020868245.1_ASM2086824v1_c...</td>\n",
       "      <td>ASM2086824v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM20868...</td>\n",
       "      <td>outbreak_genome/GCA_020868245.1_ASM2086824v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>outbreak_genome/GCA_003345275.1_ASM334527v1_cd...</td>\n",
       "      <td>ASM334527v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM33452...</td>\n",
       "      <td>outbreak_genome/GCA_003345275.1_ASM334527v1_cd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>outbreak_genome/GCA_026000835.1_ASM2600083v1_c...</td>\n",
       "      <td>ASM2600083v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM26000...</td>\n",
       "      <td>outbreak_genome/GCA_026000835.1_ASM2600083v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>outbreak_genome/GCA_027943905.1_ASM2794390v1_c...</td>\n",
       "      <td>ASM2794390v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM27943...</td>\n",
       "      <td>outbreak_genome/GCA_027943905.1_ASM2794390v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>outbreak_genome/GCA_032941915.1_ASM3294191v1_c...</td>\n",
       "      <td>ASM3294191v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM32941...</td>\n",
       "      <td>outbreak_genome/GCA_032941915.1_ASM3294191v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>outbreak_genome/GCA_037224975.1_ASM3722497v1_c...</td>\n",
       "      <td>ASM3722497v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM37224...</td>\n",
       "      <td>outbreak_genome/GCA_037224975.1_ASM3722497v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>outbreak_genome/GCA_037227285.1_ASM3722728v1_c...</td>\n",
       "      <td>ASM3722728v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM37227...</td>\n",
       "      <td>outbreak_genome/GCA_037227285.1_ASM3722728v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>outbreak_genome/GCA_037287405.1_ASM3728740v1_c...</td>\n",
       "      <td>ASM3728740v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM37287...</td>\n",
       "      <td>outbreak_genome/GCA_037287405.1_ASM3728740v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>outbreak_genome/GCA_037287515.1_ASM3728751v1_c...</td>\n",
       "      <td>ASM3728751v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM37287...</td>\n",
       "      <td>outbreak_genome/GCA_037287515.1_ASM3728751v1_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>outbreak_genome/GCA_037287535.1_ASM3728753v1_c...</td>\n",
       "      <td>ASM3728753v1</td>\n",
       "      <td>results/HGT_DB/analysis/card/outbreak_ASM37287...</td>\n",
       "      <td>outbreak_genome/GCA_037287535.1_ASM3728753v1_c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            file_path      asm_name   \n",
       "0   outbreak_genome/GCA_013459015.1_ASM1345901v1_c...  ASM1345901v1  \\\n",
       "1   outbreak_genome/GCA_013459175.1_ASM1345917v1_c...  ASM1345917v1   \n",
       "2   outbreak_genome/GCA_015209985.1_ASM1520998v1_c...  ASM1520998v1   \n",
       "3   outbreak_genome/GCA_020868245.1_ASM2086824v1_c...  ASM2086824v1   \n",
       "4   outbreak_genome/GCA_003345275.1_ASM334527v1_cd...   ASM334527v1   \n",
       "5   outbreak_genome/GCA_026000835.1_ASM2600083v1_c...  ASM2600083v1   \n",
       "6   outbreak_genome/GCA_027943905.1_ASM2794390v1_c...  ASM2794390v1   \n",
       "7   outbreak_genome/GCA_032941915.1_ASM3294191v1_c...  ASM3294191v1   \n",
       "8   outbreak_genome/GCA_037224975.1_ASM3722497v1_c...  ASM3722497v1   \n",
       "9   outbreak_genome/GCA_037227285.1_ASM3722728v1_c...  ASM3722728v1   \n",
       "10  outbreak_genome/GCA_037287405.1_ASM3728740v1_c...  ASM3728740v1   \n",
       "11  outbreak_genome/GCA_037287515.1_ASM3728751v1_c...  ASM3728751v1   \n",
       "12  outbreak_genome/GCA_037287535.1_ASM3728753v1_c...  ASM3728753v1   \n",
       "\n",
       "                                    arg_folder_target   \n",
       "0   results/HGT_DB/analysis/card/outbreak_ASM13459...  \\\n",
       "1   results/HGT_DB/analysis/card/outbreak_ASM13459...   \n",
       "2   results/HGT_DB/analysis/card/outbreak_ASM15209...   \n",
       "3   results/HGT_DB/analysis/card/outbreak_ASM20868...   \n",
       "4   results/HGT_DB/analysis/card/outbreak_ASM33452...   \n",
       "5   results/HGT_DB/analysis/card/outbreak_ASM26000...   \n",
       "6   results/HGT_DB/analysis/card/outbreak_ASM27943...   \n",
       "7   results/HGT_DB/analysis/card/outbreak_ASM32941...   \n",
       "8   results/HGT_DB/analysis/card/outbreak_ASM37224...   \n",
       "9   results/HGT_DB/analysis/card/outbreak_ASM37227...   \n",
       "10  results/HGT_DB/analysis/card/outbreak_ASM37287...   \n",
       "11  results/HGT_DB/analysis/card/outbreak_ASM37287...   \n",
       "12  results/HGT_DB/analysis/card/outbreak_ASM37287...   \n",
       "\n",
       "                                        file_path_csv  \n",
       "0   outbreak_genome/GCA_013459015.1_ASM1345901v1_c...  \n",
       "1   outbreak_genome/GCA_013459175.1_ASM1345917v1_c...  \n",
       "2   outbreak_genome/GCA_015209985.1_ASM1520998v1_c...  \n",
       "3   outbreak_genome/GCA_020868245.1_ASM2086824v1_c...  \n",
       "4   outbreak_genome/GCA_003345275.1_ASM334527v1_cd...  \n",
       "5   outbreak_genome/GCA_026000835.1_ASM2600083v1_c...  \n",
       "6   outbreak_genome/GCA_027943905.1_ASM2794390v1_c...  \n",
       "7   outbreak_genome/GCA_032941915.1_ASM3294191v1_c...  \n",
       "8   outbreak_genome/GCA_037224975.1_ASM3722497v1_c...  \n",
       "9   outbreak_genome/GCA_037227285.1_ASM3722728v1_c...  \n",
       "10  outbreak_genome/GCA_037287405.1_ASM3728740v1_c...  \n",
       "11  outbreak_genome/GCA_037287515.1_ASM3728751v1_c...  \n",
       "12  outbreak_genome/GCA_037287535.1_ASM3728753v1_c...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outbreak_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ASM1345901v1\n",
      "Total ARG 55\n",
      "Total extraneous counted 427\n",
      "Toal ARG that is extraneous 6\n",
      "ASM1345917v1\n",
      "Total ARG 61\n",
      "Total extraneous counted 468\n",
      "Toal ARG that is extraneous 9\n",
      "ASM1520998v1\n",
      "Total ARG 8\n",
      "Total extraneous counted 544\n",
      "Toal ARG that is extraneous 0\n",
      "ASM2086824v1\n",
      "Total ARG 82\n",
      "Total extraneous counted 534\n",
      "Toal ARG that is extraneous 18\n",
      "ASM334527v1\n",
      "Total ARG 9\n",
      "Total extraneous counted 395\n",
      "Toal ARG that is extraneous 3\n",
      "ASM2600083v1\n",
      "Total ARG 28\n",
      "Total extraneous counted 477\n",
      "Toal ARG that is extraneous 3\n",
      "ASM2794390v1\n",
      "Total ARG 12\n",
      "Total extraneous counted 353\n",
      "Toal ARG that is extraneous 2\n",
      "ASM3294191v1\n",
      "Total ARG 60\n",
      "Total extraneous counted 498\n",
      "Toal ARG that is extraneous 1\n",
      "ASM3722497v1\n",
      "Total ARG 34\n",
      "Total extraneous counted 542\n",
      "Toal ARG that is extraneous 2\n",
      "ASM3722728v1\n",
      "Total ARG 17\n",
      "Total extraneous counted 454\n",
      "Toal ARG that is extraneous 1\n",
      "ASM3728740v1\n",
      "Total ARG 60\n",
      "Total extraneous counted 590\n",
      "Toal ARG that is extraneous 2\n",
      "ASM3728751v1\n",
      "Total ARG 47\n",
      "Total extraneous counted 562\n",
      "Toal ARG that is extraneous 2\n",
      "ASM3728753v1\n",
      "Total ARG 61\n",
      "Total extraneous counted 537\n",
      "Toal ARG that is extraneous 1\n"
     ]
    }
   ],
   "source": [
    "to_csv_3 = []\n",
    "for i in range(len(outbreak_list)):\n",
    "    df = pd.read_csv(outbreak_list.iloc[i]['file_path_csv'],index_col=0)\n",
    "    arg_genes_json = open_result_json(outbreak_list.iloc[i]['arg_folder_target'])\n",
    "    \n",
    "    arg_genes = list(arg_genes_json.keys())\n",
    "    \n",
    "    df['extraneous'] = [check_extraneity(df.iloc[i]) for i in range(len(df))]\n",
    "    \n",
    "    print(f\"{outbreak_list.iloc[i]['asm_name']}\")\n",
    "    print(f\"Total ARG {len(arg_genes)}\")\n",
    "    print(f\"Total extraneous counted {df[df['extraneous'] == True]['extraneous'].count()}\")    \n",
    "    arg_genes_df = df.loc[arg_genes]\n",
    "    print(f\"Toal ARG that is extraneous {arg_genes_df[arg_genes_df['extraneous'] == True]['extraneous'].count()}\")\n",
    "    \n",
    "    to_csv_3.append((outbreak_list.iloc[i]['asm_name'],len(df), len(arg_genes), df[df['extraneous'] == True]['extraneous'].count(), arg_genes_df[arg_genes_df['extraneous'] == True]['extraneous'].count()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('ASM1345901v1', 4473, 55, 427, 6),\n",
       " ('ASM1345917v1', 4723, 61, 468, 9),\n",
       " ('ASM1520998v1', 5130, 8, 544, 0),\n",
       " ('ASM2086824v1', 5016, 82, 534, 18),\n",
       " ('ASM334527v1', 3278, 9, 395, 3),\n",
       " ('ASM2600083v1', 4993, 28, 477, 3),\n",
       " ('ASM2794390v1', 2645, 12, 353, 2),\n",
       " ('ASM3294191v1', 5786, 60, 498, 1),\n",
       " ('ASM3722497v1', 5352, 34, 542, 2),\n",
       " ('ASM3722728v1', 4933, 17, 454, 1),\n",
       " ('ASM3728740v1', 6408, 60, 590, 2),\n",
       " ('ASM3728751v1', 5675, 47, 562, 2),\n",
       " ('ASM3728753v1', 6483, 61, 537, 1)]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_csv_3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = 'ignore/data_statistic_report_thesis_13.csv'\n",
    "\n",
    "\n",
    "with open(file_name,'w') as out:\n",
    "    csv_out=csv.writer(out)\n",
    "    #csv_out.writerow(['genome','type','gene','SD1','SD2','SD3','SDT'])\n",
    "    csv_out.writerow(['genome','len','ARG','extraneous count', 'overlap'])\n",
    "    csv_out.writerows(to_csv_3)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ignore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group 1: [18.6642269, 10.26252983, 7.476635514, 5.806451613, 5.075845974, 4.384672071, 3.353293413], Sum: 55.023655315, Average: 7.860522187857143\n",
      "Group 2: [16.5582068, 10.34883721, 7.528957529, 6.076759062, 5.098743268, 4.548286604, 3.58056266, 1.102941176], Sum: 54.843294309, Average: 6.855411788625\n",
      "Group 3: [15.75973211, 10.44133477, 8.014142605, 6.219312602, 5.179982441, 4.481792717, 3.612523415, 1.175088132], Sum: 54.883908792, Average: 6.860488599\n",
      "Group 4: [13.42738993, 10.53719008, 8.395696913, 6.816551064, 5.660377358, 4.705039408, 4.047802621, 3.272450533], Sum: 56.862497907, Average: 7.107812238375\n",
      "Group 5: [13.25347759, 10.63011063, 8.509485095, 6.662456584, 5.730937348, 4.946043165, 3.970004411, 1.320132013], Sum: 55.02264683599999, Average: 6.877830854499999\n",
      "Group 6: [12.71565495, 10.76716016, 10.02022018, 6.465214336, 5.357677342, 4.578155657, 3.930131004], Sum: 53.834213629000004, Average: 7.690601947\n",
      "Group 7: [12.53697383, 11.13824192, 9.975786925, 6.25, 5.5028463, 4.563758389, 3.783078643, 0], Sum: 53.750686007000006, Average: 6.718835750875001\n",
      "Group 8: [12.32091691, 11.49601471, 8.953453899, 6.586270872, 5.514639842, 4.583723106, 4.224422442, 1.935171746], Sum: 55.614613527, Average: 6.951826690875\n",
      "Group 9: [12.19799646, 11.65827271, 8.533916849, 6.809210526, 5.525606469, 4.977203647, 3.976034858, 1.954397394], Sum: 55.632638913, Average: 6.954079864125\n",
      "Group 10: [12.18795888, 11.85612789, 8.133848134, 7.129731716, 5.520304569, 4.773561812, 4.022346369, 2.819548872], Sum: 56.44342824199999, Average: 7.055428530249999\n"
     ]
    }
   ],
   "source": [
    "def balanced_grouping(numbers, num_groups=10):\n",
    "    # Sort numbers in descending order\n",
    "    sorted_numbers = sorted(numbers, reverse=True)\n",
    "    \n",
    "    # Initialize groups and their sums\n",
    "    groups = [[] for _ in range(num_groups)]\n",
    "    group_sums = [0] * num_groups\n",
    "    \n",
    "    # Distribute numbers using the greedy algorithm\n",
    "    for number in sorted_numbers:\n",
    "        min_index = group_sums.index(min(group_sums))\n",
    "        groups[min_index].append(number)\n",
    "        group_sums[min_index] += number\n",
    "    \n",
    "    return groups\n",
    "\n",
    "# Example usage\n",
    "#numbers = [15, 13, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]\n",
    "numbers= [18.6642269,\n",
    "16.5582068,\n",
    "15.75973211,\n",
    "13.42738993,\n",
    "13.25347759,\n",
    "12.71565495,\n",
    "12.53697383,\n",
    "12.32091691,\n",
    "12.19799646,\n",
    "12.18795888,\n",
    "11.85612789,\n",
    "11.65827271,\n",
    "11.49601471,\n",
    "11.13824192,\n",
    "10.76716016,\n",
    "10.63011063,\n",
    "10.53719008,\n",
    "10.44133477,\n",
    "10.34883721,\n",
    "10.26252983,\n",
    "10.02022018,\n",
    "9.975786925,\n",
    "8.953453899,\n",
    "8.533916849,\n",
    "8.509485095,\n",
    "8.395696913,\n",
    "8.133848134,\n",
    "8.014142605,\n",
    "7.528957529,\n",
    "7.476635514,\n",
    "7.129731716,\n",
    "6.816551064,\n",
    "6.809210526,\n",
    "6.662456584,\n",
    "6.586270872,\n",
    "6.465214336,\n",
    "6.25,\n",
    "6.219312602,\n",
    "6.076759062,\n",
    "5.806451613,\n",
    "5.730937348,\n",
    "5.660377358,\n",
    "5.525606469,\n",
    "5.520304569,\n",
    "5.514639842,\n",
    "5.5028463,\n",
    "5.357677342,\n",
    "5.179982441,\n",
    "5.098743268,\n",
    "5.075845974,\n",
    "4.977203647,\n",
    "4.946043165,\n",
    "4.773561812,\n",
    "4.705039408,\n",
    "4.583723106,\n",
    "4.578155657,\n",
    "4.563758389,\n",
    "4.548286604,\n",
    "4.481792717,\n",
    "4.384672071,\n",
    "4.224422442,\n",
    "4.047802621,\n",
    "4.022346369,\n",
    "3.976034858,\n",
    "3.970004411,\n",
    "3.930131004,\n",
    "3.783078643,\n",
    "3.612523415,\n",
    "3.58056266,\n",
    "3.353293413,\n",
    "3.272450533,\n",
    "2.819548872,\n",
    "1.954397394,\n",
    "1.935171746,\n",
    "1.320132013,\n",
    "1.175088132,\n",
    "1.102941176,\n",
    "0]\n",
    "\n",
    "groups = balanced_grouping(numbers)\n",
    "\n",
    "for i, group in enumerate(groups):\n",
    "    print(f\"Group {i+1}: {group}, Sum: {sum(group)}, Average: {sum(group)/len(group) if len(group) > 0 else 0}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group 1: Indices ['vcolc2.csv', 'nmen2.csv', 'cglu.csv', 'hpyl.csv', 'cpneu3.csv', 'rconorii.csv'] Average: 8.247230505000001\n",
      "Group 2: Indices ['xfas.csv', 'styp.csv', 'xcamp.csv', 'linno.csv', 'mtub2.csv', 'llac.csv', 'bbur.csv', 'baphi.csv'] Average: 6.234207205000001\n",
      "Group 3: Indices ['btheta.csv', 'rsolac2.csv', 'atum1c1.csv', 'mpneu.csv', 'cjen.csv', 'caceto.csv', 'cmur.csv', 'bsp.csv'] Average: 6.110262462124999\n",
      "Group 4: Indices ['bsub.csv', 'rsola.csv', 'bhal.csv', 'xcitri.csv', 'bmelic2.csv', 'mtub.csv', 'dra1.csv'] Average: 7.253337030999999\n",
      "Group 5: Indices ['tteng.csv', 'tmar.csv', 'tpal.csv', 'synecho.csv', 'atum2c1.csv', 'hpyl99.csv', 'fnucl.csv'] Average: 7.069128553714286\n",
      "Group 6: Indices ['vvul1c2.csv', 'mgen.csv', 'spyo2.csv', 'pmul.csv', 'paer.csv', 'scoel.csv', 'mpul.csv'] Average: 7.269772383142857\n",
      "Group 7: Indices ['sent.csv', 'nmen1.csv', 'vcolc1.csv', 'atum1c2.csv', 'hinf.csv', 'mlep.csv', 'uure.csv'] Average: 7.071263411857141\n",
      "Group 8: Indices ['spneu1.csv', 'vvul1c1.csv', 'mloti.csv', 'bmelic1.csv', 'cpneu.csv', 'ctra.csv', 'ccres.csv'] Average: 7.272883485857141\n",
      "Group 9: Indices ['vvul2c2.csv', 'ecoli3.csv', 'ecoli.csv', 'atum2c2.csv', 'smel.csv', 'sau1.csv', 'nsp.csv'] Average: 7.28532185157143\n",
      "Group 10: Indices ['spneu2.csv', 'vvul2c1.csv', 'spyo.csv', 'dra2.csv', 'sau3.csv', 'aquae.csv', 'cperf.csv'] Average: 7.189003481571427\n",
      "Group 11: Indices ['ctep.csv', 'ecoli2.csv', 'ypestis.csv', 'lmono.csv', 'cpneu2.csv', 'sau2.csv', 'rpxx.csv'] Average: 7.256781674142857\n"
     ]
    }
   ],
   "source": [
    "def balanced_grouping_with_indices(numbers, num_groups=11):\n",
    "    # Pair each number with its original index and sort in descending order by number\n",
    "    indexed_numbers = sorted(enumerate(numbers), key=lambda x: x[1], reverse=True)\n",
    "    \n",
    "    # Initialize groups and their sums\n",
    "    groups = [[] for _ in range(num_groups)]\n",
    "    group_sums = [0] * num_groups\n",
    "    \n",
    "    # Distribute numbers using the greedy algorithm\n",
    "    for index, number in indexed_numbers:\n",
    "        min_index = group_sums.index(min(group_sums))\n",
    "        groups[min_index].append((index, number))\n",
    "        group_sums[min_index] += number\n",
    "    \n",
    "    return groups\n",
    "\n",
    "# Example usage\n",
    "numbers= [18.6642269,\n",
    "16.5582068,\n",
    "15.75973211,\n",
    "13.42738993,\n",
    "13.25347759,\n",
    "12.71565495,\n",
    "12.53697383,\n",
    "12.32091691,\n",
    "12.19799646,\n",
    "12.18795888,\n",
    "11.85612789,\n",
    "11.65827271,\n",
    "11.49601471,\n",
    "11.13824192,\n",
    "10.76716016,\n",
    "10.63011063,\n",
    "10.53719008,\n",
    "10.44133477,\n",
    "10.34883721,\n",
    "10.26252983,\n",
    "10.02022018,\n",
    "9.975786925,\n",
    "8.953453899,\n",
    "8.533916849,\n",
    "8.509485095,\n",
    "8.395696913,\n",
    "8.133848134,\n",
    "8.014142605,\n",
    "7.528957529,\n",
    "7.476635514,\n",
    "7.129731716,\n",
    "6.816551064,\n",
    "6.809210526,\n",
    "6.662456584,\n",
    "6.586270872,\n",
    "6.465214336,\n",
    "6.25,\n",
    "6.219312602,\n",
    "6.076759062,\n",
    "5.806451613,\n",
    "5.730937348,\n",
    "5.660377358,\n",
    "5.525606469,\n",
    "5.520304569,\n",
    "5.514639842,\n",
    "5.5028463,\n",
    "5.357677342,\n",
    "5.179982441,\n",
    "5.098743268,\n",
    "5.075845974,\n",
    "4.977203647,\n",
    "4.946043165,\n",
    "4.773561812,\n",
    "4.705039408,\n",
    "4.583723106,\n",
    "4.578155657,\n",
    "4.563758389,\n",
    "4.548286604,\n",
    "4.481792717,\n",
    "4.384672071,\n",
    "4.224422442,\n",
    "4.047802621,\n",
    "4.022346369,\n",
    "3.976034858,\n",
    "3.970004411,\n",
    "3.930131004,\n",
    "3.783078643,\n",
    "3.612523415,\n",
    "3.58056266,\n",
    "3.353293413,\n",
    "3.272450533,\n",
    "2.819548872,\n",
    "1.954397394,\n",
    "1.935171746,\n",
    "1.320132013,\n",
    "1.175088132,\n",
    "1.102941176,\n",
    "0]\n",
    "coressponding_files = [\n",
    "'vcolc2.csv',\n",
    "'xfas.csv',\n",
    "'btheta.csv',\n",
    "'bsub.csv',\n",
    "'tteng.csv',\n",
    "'vvul1c2.csv',\n",
    "'sent.csv',\n",
    "'spneu1.csv',\n",
    "'vvul2c2.csv',\n",
    "'spneu2.csv',\n",
    "'ctep.csv',\n",
    "'ecoli2.csv',\n",
    "'vvul2c1.csv',\n",
    "'ecoli3.csv',\n",
    "'vvul1c1.csv',\n",
    "'nmen1.csv',\n",
    "'mgen.csv',\n",
    "'tmar.csv',\n",
    "'rsola.csv',\n",
    "'rsolac2.csv',\n",
    "'styp.csv',\n",
    "'nmen2.csv',\n",
    "'mloti.csv',\n",
    "'vcolc1.csv',\n",
    "'spyo2.csv',\n",
    "'ecoli.csv',\n",
    "'ypestis.csv',\n",
    "'spyo.csv',\n",
    "'tpal.csv',\n",
    "'bhal.csv',\n",
    "'atum1c1.csv',\n",
    "'xcamp.csv',\n",
    "'cglu.csv',\n",
    "'synecho.csv',\n",
    "'xcitri.csv',\n",
    "'lmono.csv',\n",
    "'dra2.csv',\n",
    "'atum1c2.csv',\n",
    "'atum2c2.csv',\n",
    "'pmul.csv',\n",
    "'bmelic1.csv',\n",
    "'mpneu.csv',\n",
    "'linno.csv',\n",
    "'hpyl.csv',\n",
    "'paer.csv',\n",
    "'cpneu.csv',\n",
    "'smel.csv',\n",
    "'bmelic2.csv',\n",
    "'atum2c1.csv',\n",
    "'hinf.csv',\n",
    "'sau3.csv',\n",
    "'cpneu2.csv',\n",
    "'cjen.csv',\n",
    "'mtub2.csv',\n",
    "'cpneu3.csv',\n",
    "'aquae.csv',\n",
    "'hpyl99.csv',\n",
    "'mlep.csv',\n",
    "'mtub.csv',\n",
    "'sau2.csv',\n",
    "'scoel.csv',\n",
    "'sau1.csv',\n",
    "'ctra.csv',\n",
    "'caceto.csv',\n",
    "'llac.csv',\n",
    "'rconorii.csv',\n",
    "'nsp.csv',\n",
    "'ccres.csv',\n",
    "'mpul.csv',\n",
    "'rpxx.csv',\n",
    "'dra1.csv',\n",
    "'cperf.csv',\n",
    "'uure.csv',\n",
    "'fnucl.csv',\n",
    "'cmur.csv',\n",
    "'bbur.csv',\n",
    "'baphi.csv',\n",
    "'bsp.csv',\n",
    "]\n",
    "\n",
    "groups = balanced_grouping_with_indices(numbers)\n",
    "\n",
    "for i, group in enumerate(groups):\n",
    "    indices = [idx for idx, num in group]\n",
    "    values = [num for idx, num in group]\n",
    "    file_group = [coressponding_files[i] for i in indices ]\n",
    "    #print(f\"Group {i+1}: Indices {file_group}, Values {values}, Sum: {sum(values)}, Average: {sum(values)/len(values) if len(values) > 0 else 0}\")\n",
    "    print(f\"Group {i+1}: Indices {file_group} Average: {sum(values)/len(values) if len(values) > 0 else 0}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "list indices must be integers or slices, not list",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[84], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m coressponding_files[[\u001b[39m0\u001b[39;49m,\u001b[39m1\u001b[39;49m,\u001b[39m2\u001b[39;49m]]\n",
      "\u001b[0;31mTypeError\u001b[0m: list indices must be integers or slices, not list"
     ]
    }
   ],
   "source": [
    "coressponding_files[[0,1,2]]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "eda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
