{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GC content Deviation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data.utils.NCBI.data_loader import NCBIDataLoader\n",
    "# from data.utils.HGTREE.data_loader import HGTREEDataLoader\n",
    "# test out combination\n",
    "import math \n",
    "from Bio.SeqUtils import GC123\n",
    "\n",
    "\n",
    "aquifex_aeolicus_VF5 = NCBIDataLoader('AE000657')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GC_Content_Deviation(genome, return_genomic_strips = False):\n",
    "    # Init list of extraneous genes\n",
    "    list_of_extraneous_genes = []\n",
    "    \n",
    "    # part 1: GCT, GC1, GC3 deviation\n",
    "    for gene in genome.genes:\n",
    "        dev_GCT = genome[gene]['GCT'] - genome.mean_GCT\n",
    "        \n",
    "        dev_GC1 = genome[gene]['GC1'] - genome.mean_GC1\n",
    "        dev_GC3 = genome[gene]['GC3'] - genome.mean_GC3\n",
    "        equal_sign_check = dev_GC1*dev_GC3\n",
    "        \n",
    "        # consider only more than 300bp\n",
    "        # either gct > 1.5\n",
    "        # or sign gc1 and gc3 equal and one of is > 1.5\n",
    "        if len(genome[gene]['sequence']) > 300:\n",
    "            if abs(dev_GCT) > (1.5*genome.std_GCT):\n",
    "                #print(f'HGT found at {i}')\n",
    "                list_of_extraneous_genes.append(gene)\n",
    "            elif (equal_sign_check > 0):\n",
    "                if abs(dev_GC1) > (1.5*genome.std_GC1):\n",
    "                    #print(f'HGT found at {i}')\n",
    "                    list_of_extraneous_genes.append(gene)\n",
    "                elif abs(dev_GC3) > (1.5*genome.std_GC3):\n",
    "                    #print(f'HGT found at {i}')\n",
    "                    list_of_extraneous_genes.append(gene)\n",
    "                else:\n",
    "                    pass\n",
    "            else:\n",
    "                pass\n",
    "    \n",
    "    # part two: 11 genes window\n",
    "    list_of_genes = list(genome.genes.keys())\n",
    "    genomic_strips = []\n",
    "    \n",
    "    for k in range(len(list_of_genes)-10):\n",
    "        window = {}\n",
    "        j = 0\n",
    "        while j < 11:\n",
    "            # get window\n",
    "            locust_tag = list_of_genes[k + j]\n",
    "            \n",
    "            # take genes that are more than 300bp\n",
    "            if len(genome[locust_tag]['sequence'])>300:\n",
    "                data = genome[locust_tag]\n",
    "                window[locust_tag] = data\n",
    "            # iterate        \n",
    "            j+=1\n",
    "            \n",
    "        # count total extraneous genes in window\n",
    "        extraneous_counter = 0\n",
    "        for l in window.keys():\n",
    "            if l in list_of_extraneous_genes:\n",
    "                extraneous_counter+=1\n",
    "        \n",
    "        # check windows with more than or\n",
    "        # equal to 5 extraneous genes\n",
    "        if extraneous_counter >=5:\n",
    "            # add their sequences together\n",
    "            window_sequences = ''\n",
    "            for m in window.keys():\n",
    "                window_sequences += window[m]['sequence']\n",
    "            \n",
    "            # get standard deviation of strip\n",
    "            GCT, GC1, GC2, GC3 = GC123(window_sequences)\n",
    "            SDT = (GCT - genome.mean_GCT)/genome.std_GCT\n",
    "            SD1 = (GC1- genome.mean_GC1)/genome.std_GC1\n",
    "            SD2 = (GC2 - genome.mean_GC2)/genome.std_GC2\n",
    "            SD3 = (GC3 - genome.mean_GC3)/genome.std_GC3\n",
    "            \n",
    "            # tag genes as extraneous if they have equal deviation to the its strip\n",
    "            for n in window.keys():\n",
    "                # check only genes not in current list\n",
    "                if n not in list_of_extraneous_genes:\n",
    "                    check_SDT = window[n]['SDT']*SDT\n",
    "                    check_SD1 = window[n]['SD1']*SD1\n",
    "                    check_SD2 = window[n]['SD2']*SD2\n",
    "                    check_SD3 = window[n]['SD3']*SD3\n",
    "                    \n",
    "                    if (check_SDT > 0) and (check_SD1 > 0) and (check_SD2 > 0) and (check_SD3 > 0):\n",
    "                        list_of_extraneous_genes.append(n)\n",
    "                \n",
    "            \n",
    "            genomic_strips.append(window)\n",
    "    \n",
    "    if return_genomic_strips:\n",
    "        return list_of_extraneous_genes, genomic_strips\n",
    "    else:\n",
    "        return list_of_extraneous_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "261\n",
      "['aq_005', 'aq_013', 'aq_018', 'aq_032', 'aq_035', 'aq_050', 'aq_059', 'aq_062', 'aq_081', 'aq_082', 'aq_118', 'aq_125', 'aq_132', 'aq_141', 'aq_156', 'aq_171', 'aq_172', 'aq_182', 'aq_183', 'aq_200', 'aq_220', 'aq_221', 'aq_254', 'aq_255', 'aq_264', 'aq_293', 'aq_326', 'aq_340', 'aq_351', 'aq_359', 'aq_369', 'aq_372', 'aq_376', 'aq_377', 'aq_378', 'aq_380', 'aq_381', 'aq_382', 'aq_383', 'aq_384', 'aq_385', 'aq_386', 'aq_387', 'aq_388', 'aq_402', 'aq_407', 'aq_418', 'aq_434', 'aq_451', 'aq_465', 'aq_473', 'aq_476', 'aq_488', 'aq_503', 'aq_504', 'aq_505', 'aq_506', 'aq_507', 'aq_509', 'aq_510', 'aq_511', 'aq_515', 'aq_516', 'aq_518', 'aq_519', 'aq_520', 'aq_522', 'aq_531', 'aq_558', 'aq_573', 'aq_585', 'aq_599', 'aq_626', 'aq_629', 'aq_645', 'aq_647', 'aq_652', 'aq_662', 'aq_666', 'aq_673', 'aq_678', 'aq_706', 'aq_711', 'aq_734', 'aq_737', 'aq_740', 'aq_771', 'aq_780', 'aq_782', 'aq_812', 'aq_820', 'aq_821', 'aq_833', 'aq_837', 'aq_850', 'aq_862', 'aq_895', 'aq_908', 'aq_917', 'aq_920', 'aq_926', 'aq_940', 'aq_944', 'aq_946', 'aq_950', 'aq_993', 'aq_996', 'aq_1008', 'aq_1012', 'aq_1017', 'aq_1018', 'aq_1019', 'aq_1029', 'aq_1030', 'aq_1031', 'aq_1033', 'aq_1035', 'aq_1037', 'aq_1038', 'aq_1049', 'aq_1051', 'aq_1052', 'aq_1053', 'aq_1057', 'aq_1070', 'aq_1073', 'aq_1078', 'aq_1079', 'aq_1080', 'aq_1095', 'aq_1096', 'aq_1097', 'aq_1105', 'aq_1108', 'aq_1133', 'aq_1177', 'aq_1186', 'aq_1188', 'aq_1201', 'aq_1223', 'aq_1232', 'aq_1236', 'aq_1237', 'aq_1249', 'aq_1264', 'aq_1340', 'aq_1345', 'aq_1366', 'aq_1373', 'aq_1392', 'aq_1410', 'aq_1413', 'aq_1427', 'aq_1434', 'aq_1435', 'aq_1436', 'aq_1439', 'aq_1446', 'aq_1449', 'aq_1485', 'aq_1495', 'aq_1501', 'aq_1502', 'aq_1516', 'aq_1525', 'aq_1549', 'aq_1554', 'aq_1570', 'aq_1581', 'aq_1590', 'aq_1596', 'aq_1620', 'aq_1622', 'aq_1641', 'aq_1648', 'aq_1663', 'aq_1671', 'aq_1689', 'aq_1695', 'aq_1695a', 'aq_1706', 'aq_1719', 'aq_1732', 'aq_1742', 'aq_1768', 'aq_1771', 'aq_1780', 'aq_1784', 'aq_1793', 'aq_1795', 'aq_1811', 'aq_1827a', 'aq_1832', 'aq_1834', 'aq_1849', 'aq_1859', 'aq_1869', 'aq_1908', 'aq_1928', 'aq_1933', 'aq_1937', 'aq_1960', 'aq_1993', 'aq_2000', 'aq_2002', 'aq_2020', 'aq_2036', 'aq_2038', 'aq_2042', 'aq_2054', 'aq_2056', 'aq_2086', 'aq_2096', 'aq_2103', 'aq_2106', 'aq_2132', 'aq_2135', 'aq_2145', 'aq_2154', 'aq_2157', 'aq_2170', 'aq_2172', 'aq_2203', 'aq_aa03', 'aq_aa04', 'aq_aa05', 'aq_aa07', 'aq_aa19', 'aq_aa20', 'aq_aa21', 'aq_aa24', 'aq_aa25', 'aq_aa26', 'aq_aa27', 'aq_aa28', 'aq_aa29', 'aq_aa31', 'aq_aa32', 'aq_aa34', 'aq_aa35', 'aq_aa37', 'aq_aa38', 'aq_aa39', 'aq_367', 'aq_371', 'aq_373', 'aq_374', 'aq_497', 'aq_512', 'aq_521', 'aq_523', 'aq_1005', 'aq_1015', 'aq_1023', 'aq_1036', 'aq_1058', 'aq_1059', 'aq_1429', 'aq_aa13', 'aq_aa22', 'aq_aa23']\n"
     ]
    }
   ],
   "source": [
    "hgt_candidates=GC_Content_Deviation(aquifex_aeolicus_VF5)\n",
    "print(len(hgt_candidates))\n",
    "print(hgt_candidates)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AA Deviation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "CODE = {\n",
    "    'ttt': 'F', 'tct': 'S', 'tat': 'Y', 'tgt': 'C',\n",
    "    'ttc': 'F', 'tcc': 'S', 'tac': 'Y', 'tgc': 'C',\n",
    "    'tta': 'L', 'tca': 'S', 'taa': '*', 'tga': '*',\n",
    "    'ttg': 'L', 'tcg': 'S', 'tag': '*', 'tgg': 'W',\n",
    "    'ctt': 'L', 'cct': 'P', 'cat': 'H', 'cgt': 'R',\n",
    "    'ctc': 'L', 'ccc': 'P', 'cac': 'H', 'cgc': 'R',\n",
    "    'cta': 'L', 'cca': 'P', 'caa': 'Q', 'cga': 'R',\n",
    "    'ctg': 'L', 'ccg': 'P', 'cag': 'Q', 'cgg': 'R',\n",
    "    'att': 'I', 'act': 'T', 'aat': 'N', 'agt': 'S',\n",
    "    'atc': 'I', 'acc': 'T', 'aac': 'N', 'agc': 'S',\n",
    "    'ata': 'I', 'aca': 'T', 'aaa': 'K', 'aga': 'R',\n",
    "    'atg': 'M', 'acg': 'T', 'aag': 'K', 'agg': 'R',\n",
    "    'gtt': 'V', 'gct': 'A', 'gat': 'D', 'ggt': 'G',\n",
    "    'gtc': 'V', 'gcc': 'A', 'gac': 'D', 'ggc': 'G',\n",
    "    'gta': 'V', 'gca': 'A', 'gaa': 'E', 'gga': 'G',\n",
    "    'gtg': 'V', 'gcg': 'A', 'gag': 'E', 'ggg': 'G'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'K': ['AAA', 'AAG'],\n",
       " 'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
       " '*': ['TAA', 'TGA', 'TAG'],\n",
       " 'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
       " 'D': ['GAT', 'GAC'],\n",
       " 'F': ['TTT', 'TTC'],\n",
       " 'C': ['TGT', 'TGC'],\n",
       " 'Y': ['TAT', 'TAC'],\n",
       " 'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
       " 'E': ['GAA', 'GAG'],\n",
       " 'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
       " 'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
       " 'W': ['TGG'],\n",
       " 'I': ['ATT', 'ATC', 'ATA'],\n",
       " 'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
       " 'Q': ['CAA', 'CAG'],\n",
       " 'H': ['CAT', 'CAC'],\n",
       " 'M': ['ATG'],\n",
       " 'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
       " 'N': ['AAT', 'AAC'],\n",
       " 'G': ['GGT', 'GGC', 'GGA', 'GGG']}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_amino_acid = set([CODE[cds] for cds in CODE])\n",
    "list_of_amino_acid = {v: [cds.upper() for cds in CODE if CODE[cds]==v] for v in list_of_amino_acid}\n",
    "list_of_amino_acid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_amino_acid_content_genome_mean(genome):\n",
    "    amino_acid_count_genome = {amino_acid: 0 for amino_acid in list_of_amino_acid}\n",
    "    for locust_tag in aquifex_aeolicus_VF5.genes:\n",
    "        for amino_acid in list_of_amino_acid:\n",
    "            for codons in list_of_amino_acid[amino_acid]:\n",
    "                amino_acid_count_genome[amino_acid] += aquifex_aeolicus_VF5[locust_tag]['cub'][codons]\n",
    "\n",
    "    total_amino_acid_count = sum([amino_acid_count_genome[amino_acid] for amino_acid in amino_acid_count_genome ])\n",
    "    for i in amino_acid_count_genome:\n",
    "        amino_acid_count_genome[i]=amino_acid_count_genome[i]*100/total_amino_acid_count\n",
    "\n",
    "    genome.mean_AA = amino_acid_count_genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_amino_acid_content_gene_mean(genome):\n",
    "    # calculate mean\n",
    "    for locust_tag in genome.genes:\n",
    "        amino_acid_count_gene = {amino_acid: 0 for amino_acid in list_of_amino_acid}\n",
    "        for amino_acid in list_of_amino_acid:\n",
    "            for codons in list_of_amino_acid[amino_acid]:\n",
    "                amino_acid_count_gene[amino_acid]  += genome[locust_tag]['cub'][codons]\n",
    "                \n",
    "        total_amino_acid_count = sum([amino_acid_count_gene[amino_acid] for amino_acid in amino_acid_count_gene ])\n",
    "        for AA in amino_acid_count_gene:\n",
    "            amino_acid_count_gene[AA] = amino_acid_count_gene[AA]*100/total_amino_acid_count\n",
    "\n",
    "        genome[locust_tag]['AA_Content_mean']=amino_acid_count_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_amino_acid_content_genome_std(genome):\n",
    "    amino_acid_std = {amino_acid: 0 for amino_acid in list_of_amino_acid}\n",
    "    \n",
    "    for amino_acid in list_of_amino_acid:\n",
    "        AA_mean = genome.mean_AA[amino_acid]\n",
    "        sum_diff = 0\n",
    "        for locust_tag in genome.genes:\n",
    "            diff = (genome[locust_tag]['AA_Content_mean'][amino_acid]) - AA_mean\n",
    "            diff = diff*diff\n",
    "            sum_diff+=diff\n",
    "        \n",
    "        amino_acid_std[amino_acid]=math.sqrt(sum_diff/len(genome))\n",
    "            \n",
    "    genome.std_AA = amino_acid_std"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The bottom functions assign new attributes to genome and gene\n",
    "\n",
    "- genome_mean: add AA mean to genome\n",
    "- gene_mean: add AA mean to each gene\n",
    "- genome_std: add AA std to genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "calculate_amino_acid_content_genome_mean(aquifex_aeolicus_VF5)\n",
    "calculate_amino_acid_content_gene_mean(aquifex_aeolicus_VF5)\n",
    "calculate_amino_acid_content_genome_std(aquifex_aeolicus_VF5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_amino_acid_deviation(genome):\n",
    "    list_of_non_extraneous_genes_AA = []\n",
    "    for locust_tag in genome.genes:\n",
    "        for amino_acid in list_of_amino_acid:\n",
    "            devAA = genome[locust_tag]['AA_Content_mean'][amino_acid] - genome.mean_AA[amino_acid]\n",
    "            if devAA > 3*genome.std_AA[amino_acid]:\n",
    "                list_of_non_extraneous_genes_AA.append(locust_tag)\n",
    "                break\n",
    "    return list_of_non_extraneous_genes_AA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['aq_008',\n",
       " 'aq_012',\n",
       " 'aq_020',\n",
       " 'aq_022',\n",
       " 'aq_039',\n",
       " 'aq_045',\n",
       " 'aq_055',\n",
       " 'aq_063',\n",
       " 'aq_064a',\n",
       " 'aq_064b',\n",
       " 'aq_064c',\n",
       " 'aq_067',\n",
       " 'aq_069',\n",
       " 'aq_073',\n",
       " 'aq_074',\n",
       " 'aq_075',\n",
       " 'aq_075a',\n",
       " 'aq_080',\n",
       " 'aq_087',\n",
       " 'aq_088',\n",
       " 'aq_090',\n",
       " 'aq_098',\n",
       " 'aq_106',\n",
       " 'aq_108a',\n",
       " 'aq_108b',\n",
       " 'aq_124a',\n",
       " 'aq_125',\n",
       " 'aq_141',\n",
       " 'aq_156',\n",
       " 'aq_157',\n",
       " 'aq_158',\n",
       " 'aq_175a',\n",
       " 'aq_176',\n",
       " 'aq_177',\n",
       " 'aq_183a',\n",
       " 'aq_194',\n",
       " 'aq_250',\n",
       " 'aq_252',\n",
       " 'aq_254',\n",
       " 'aq_267',\n",
       " 'aq_293',\n",
       " 'aq_314',\n",
       " 'aq_355',\n",
       " 'aq_357a',\n",
       " 'aq_363',\n",
       " 'aq_380',\n",
       " 'aq_388',\n",
       " 'aq_389',\n",
       " 'aq_391',\n",
       " 'aq_392',\n",
       " 'aq_394',\n",
       " 'aq_401',\n",
       " 'aq_406',\n",
       " 'aq_419',\n",
       " 'aq_420',\n",
       " 'aq_449',\n",
       " 'aq_453',\n",
       " 'aq_481',\n",
       " 'aq_488',\n",
       " 'aq_504',\n",
       " 'aq_507',\n",
       " 'aq_509',\n",
       " 'aq_519',\n",
       " 'aq_531',\n",
       " 'aq_538',\n",
       " 'aq_539',\n",
       " 'aq_553',\n",
       " 'aq_577',\n",
       " 'aq_591',\n",
       " 'aq_615',\n",
       " 'aq_628',\n",
       " 'aq_652',\n",
       " 'aq_655',\n",
       " 'aq_665',\n",
       " 'aq_666',\n",
       " 'aq_671',\n",
       " 'aq_678',\n",
       " 'aq_702',\n",
       " 'aq_722',\n",
       " 'aq_735',\n",
       " 'aq_737',\n",
       " 'aq_754',\n",
       " 'aq_758',\n",
       " 'aq_766',\n",
       " 'aq_770',\n",
       " 'aq_780',\n",
       " 'aq_792',\n",
       " 'aq_792a',\n",
       " 'aq_808',\n",
       " 'aq_818',\n",
       " 'aq_834',\n",
       " 'aq_848',\n",
       " 'aq_854',\n",
       " 'aq_867a',\n",
       " 'aq_914',\n",
       " 'aq_917',\n",
       " 'aq_918',\n",
       " 'aq_919a',\n",
       " 'aq_945',\n",
       " 'aq_950',\n",
       " 'aq_952',\n",
       " 'aq_967',\n",
       " 'aq_978',\n",
       " 'aq_979',\n",
       " 'aq_1003',\n",
       " 'aq_1012',\n",
       " 'aq_1017',\n",
       " 'aq_1021',\n",
       " 'aq_1029',\n",
       " 'aq_1036',\n",
       " 'aq_1046',\n",
       " 'aq_1049',\n",
       " 'aq_1050',\n",
       " 'aq_1072',\n",
       " 'aq_1073',\n",
       " 'aq_1076',\n",
       " 'aq_1078',\n",
       " 'aq_1084',\n",
       " 'aq_1091',\n",
       " 'aq_1102',\n",
       " 'aq_1108',\n",
       " 'aq_1127',\n",
       " 'aq_1130',\n",
       " 'aq_1166',\n",
       " 'aq_1171a',\n",
       " 'aq_1183',\n",
       " 'aq_1184',\n",
       " 'aq_1185',\n",
       " 'aq_1186',\n",
       " 'aq_1188',\n",
       " 'aq_1192a',\n",
       " 'aq_1200',\n",
       " 'aq_1207',\n",
       " 'aq_1223',\n",
       " 'aq_1230',\n",
       " 'aq_1232',\n",
       " 'aq_1249',\n",
       " 'aq_1253',\n",
       " 'aq_1256',\n",
       " 'aq_1264',\n",
       " 'aq_1273',\n",
       " 'aq_1283',\n",
       " 'aq_1303a',\n",
       " 'aq_1317',\n",
       " 'aq_1324',\n",
       " 'aq_1327',\n",
       " 'aq_1339',\n",
       " 'aq_1345',\n",
       " 'aq_1349',\n",
       " 'aq_1374',\n",
       " 'aq_1375',\n",
       " 'aq_1378',\n",
       " 'aq_1387',\n",
       " 'aq_1391',\n",
       " 'aq_1392',\n",
       " 'aq_1401',\n",
       " 'aq_1426',\n",
       " 'aq_1432',\n",
       " 'aq_1435',\n",
       " 'aq_1439',\n",
       " 'aq_1466',\n",
       " 'aq_1476',\n",
       " 'aq_1480',\n",
       " 'aq_1484a',\n",
       " 'aq_1501',\n",
       " 'aq_1542',\n",
       " 'aq_1550',\n",
       " 'aq_1579',\n",
       " 'aq_1581',\n",
       " 'aq_1586',\n",
       " 'aq_1625',\n",
       " 'aq_1631',\n",
       " 'aq_1639',\n",
       " 'aq_1640',\n",
       " 'aq_1641a',\n",
       " 'aq_1648',\n",
       " 'aq_1649',\n",
       " 'aq_1651a',\n",
       " 'aq_1660',\n",
       " 'aq_1663',\n",
       " 'aq_1666',\n",
       " 'aq_1681',\n",
       " 'aq_1695a',\n",
       " 'aq_1714',\n",
       " 'aq_1717a',\n",
       " 'aq_1719',\n",
       " 'aq_1767',\n",
       " 'aq_1789',\n",
       " 'aq_1791',\n",
       " 'aq_1793',\n",
       " 'aq_1795',\n",
       " 'aq_1796',\n",
       " 'aq_1811',\n",
       " 'aq_1823',\n",
       " 'aq_1827a',\n",
       " 'aq_1834',\n",
       " 'aq_1854',\n",
       " 'aq_1859',\n",
       " 'aq_1894',\n",
       " 'aq_1896',\n",
       " 'aq_1916',\n",
       " 'aq_1930a',\n",
       " 'aq_1933',\n",
       " 'aq_1937',\n",
       " 'aq_1954',\n",
       " 'aq_1961',\n",
       " 'aq_1962',\n",
       " 'aq_1974',\n",
       " 'aq_1979',\n",
       " 'aq_1982',\n",
       " 'aq_1998',\n",
       " 'aq_2000',\n",
       " 'aq_2043',\n",
       " 'aq_2049',\n",
       " 'aq_2051',\n",
       " 'aq_2058',\n",
       " 'aq_2063',\n",
       " 'aq_2104',\n",
       " 'aq_2129',\n",
       " 'aq_2154',\n",
       " 'aq_2157',\n",
       " 'aq_2159',\n",
       " 'aq_2185',\n",
       " 'aq_2188',\n",
       " 'aq_2191a',\n",
       " 'aq_2195',\n",
       " 'aq_aa13',\n",
       " 'aq_aa17',\n",
       " 'aq_aa18',\n",
       " 'aq_aa24',\n",
       " 'aq_aa32',\n",
       " 'aq_aa40']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_non_extraneous_genes_AA = check_amino_acid_deviation(aquifex_aeolicus_VF5)\n",
    "list_of_non_extraneous_genes_AA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "find out extraneous genes to be taken out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58\n",
      "203\n"
     ]
    }
   ],
   "source": [
    "take_out_list = []\n",
    "list_of_hgt =[]\n",
    "for i in hgt_candidates:\n",
    "    if i in list_of_non_extraneous_genes_AA:\n",
    "        take_out_list.append(i)\n",
    "    else:\n",
    "        list_of_hgt.append(i)\n",
    "        \n",
    "\n",
    "print(len(take_out_list))\n",
    "print(len(list_of_hgt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['aq_005',\n",
       " 'aq_013',\n",
       " 'aq_018',\n",
       " 'aq_032',\n",
       " 'aq_035',\n",
       " 'aq_050',\n",
       " 'aq_059',\n",
       " 'aq_062',\n",
       " 'aq_081',\n",
       " 'aq_082',\n",
       " 'aq_118',\n",
       " 'aq_132',\n",
       " 'aq_171',\n",
       " 'aq_172',\n",
       " 'aq_182',\n",
       " 'aq_183',\n",
       " 'aq_200',\n",
       " 'aq_220',\n",
       " 'aq_221',\n",
       " 'aq_255',\n",
       " 'aq_264',\n",
       " 'aq_326',\n",
       " 'aq_340',\n",
       " 'aq_351',\n",
       " 'aq_359',\n",
       " 'aq_369',\n",
       " 'aq_372',\n",
       " 'aq_376',\n",
       " 'aq_377',\n",
       " 'aq_378',\n",
       " 'aq_381',\n",
       " 'aq_382',\n",
       " 'aq_383',\n",
       " 'aq_384',\n",
       " 'aq_385',\n",
       " 'aq_386',\n",
       " 'aq_387',\n",
       " 'aq_402',\n",
       " 'aq_407',\n",
       " 'aq_418',\n",
       " 'aq_434',\n",
       " 'aq_451',\n",
       " 'aq_465',\n",
       " 'aq_473',\n",
       " 'aq_476',\n",
       " 'aq_503',\n",
       " 'aq_505',\n",
       " 'aq_506',\n",
       " 'aq_510',\n",
       " 'aq_511',\n",
       " 'aq_515',\n",
       " 'aq_516',\n",
       " 'aq_518',\n",
       " 'aq_520',\n",
       " 'aq_522',\n",
       " 'aq_558',\n",
       " 'aq_573',\n",
       " 'aq_585',\n",
       " 'aq_599',\n",
       " 'aq_626',\n",
       " 'aq_629',\n",
       " 'aq_645',\n",
       " 'aq_647',\n",
       " 'aq_662',\n",
       " 'aq_673',\n",
       " 'aq_706',\n",
       " 'aq_711',\n",
       " 'aq_734',\n",
       " 'aq_740',\n",
       " 'aq_771',\n",
       " 'aq_782',\n",
       " 'aq_812',\n",
       " 'aq_820',\n",
       " 'aq_821',\n",
       " 'aq_833',\n",
       " 'aq_837',\n",
       " 'aq_850',\n",
       " 'aq_862',\n",
       " 'aq_895',\n",
       " 'aq_908',\n",
       " 'aq_920',\n",
       " 'aq_926',\n",
       " 'aq_940',\n",
       " 'aq_944',\n",
       " 'aq_946',\n",
       " 'aq_993',\n",
       " 'aq_996',\n",
       " 'aq_1008',\n",
       " 'aq_1018',\n",
       " 'aq_1019',\n",
       " 'aq_1030',\n",
       " 'aq_1031',\n",
       " 'aq_1033',\n",
       " 'aq_1035',\n",
       " 'aq_1037',\n",
       " 'aq_1038',\n",
       " 'aq_1051',\n",
       " 'aq_1052',\n",
       " 'aq_1053',\n",
       " 'aq_1057',\n",
       " 'aq_1070',\n",
       " 'aq_1079',\n",
       " 'aq_1080',\n",
       " 'aq_1095',\n",
       " 'aq_1096',\n",
       " 'aq_1097',\n",
       " 'aq_1105',\n",
       " 'aq_1133',\n",
       " 'aq_1177',\n",
       " 'aq_1201',\n",
       " 'aq_1236',\n",
       " 'aq_1237',\n",
       " 'aq_1340',\n",
       " 'aq_1366',\n",
       " 'aq_1373',\n",
       " 'aq_1410',\n",
       " 'aq_1413',\n",
       " 'aq_1427',\n",
       " 'aq_1434',\n",
       " 'aq_1436',\n",
       " 'aq_1446',\n",
       " 'aq_1449',\n",
       " 'aq_1485',\n",
       " 'aq_1495',\n",
       " 'aq_1502',\n",
       " 'aq_1516',\n",
       " 'aq_1525',\n",
       " 'aq_1549',\n",
       " 'aq_1554',\n",
       " 'aq_1570',\n",
       " 'aq_1590',\n",
       " 'aq_1596',\n",
       " 'aq_1620',\n",
       " 'aq_1622',\n",
       " 'aq_1641',\n",
       " 'aq_1671',\n",
       " 'aq_1689',\n",
       " 'aq_1695',\n",
       " 'aq_1706',\n",
       " 'aq_1732',\n",
       " 'aq_1742',\n",
       " 'aq_1768',\n",
       " 'aq_1771',\n",
       " 'aq_1780',\n",
       " 'aq_1784',\n",
       " 'aq_1832',\n",
       " 'aq_1849',\n",
       " 'aq_1869',\n",
       " 'aq_1908',\n",
       " 'aq_1928',\n",
       " 'aq_1960',\n",
       " 'aq_1993',\n",
       " 'aq_2002',\n",
       " 'aq_2020',\n",
       " 'aq_2036',\n",
       " 'aq_2038',\n",
       " 'aq_2042',\n",
       " 'aq_2054',\n",
       " 'aq_2056',\n",
       " 'aq_2086',\n",
       " 'aq_2096',\n",
       " 'aq_2103',\n",
       " 'aq_2106',\n",
       " 'aq_2132',\n",
       " 'aq_2135',\n",
       " 'aq_2145',\n",
       " 'aq_2170',\n",
       " 'aq_2172',\n",
       " 'aq_2203',\n",
       " 'aq_aa03',\n",
       " 'aq_aa04',\n",
       " 'aq_aa05',\n",
       " 'aq_aa07',\n",
       " 'aq_aa19',\n",
       " 'aq_aa20',\n",
       " 'aq_aa21',\n",
       " 'aq_aa25',\n",
       " 'aq_aa26',\n",
       " 'aq_aa27',\n",
       " 'aq_aa28',\n",
       " 'aq_aa29',\n",
       " 'aq_aa31',\n",
       " 'aq_aa34',\n",
       " 'aq_aa35',\n",
       " 'aq_aa37',\n",
       " 'aq_aa38',\n",
       " 'aq_aa39',\n",
       " 'aq_367',\n",
       " 'aq_371',\n",
       " 'aq_373',\n",
       " 'aq_374',\n",
       " 'aq_497',\n",
       " 'aq_512',\n",
       " 'aq_521',\n",
       " 'aq_523',\n",
       " 'aq_1005',\n",
       " 'aq_1015',\n",
       " 'aq_1023',\n",
       " 'aq_1058',\n",
       " 'aq_1059',\n",
       " 'aq_1429',\n",
       " 'aq_aa22',\n",
       " 'aq_aa23']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_hgt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'K': 9.407543465954214,\n",
       " 'P': 4.0494915226625725,\n",
       " '*': 0.3164492520475529,\n",
       " 'S': 4.7950003461800055,\n",
       " 'D': 4.304646550214428,\n",
       " 'F': 5.135274928014923,\n",
       " 'C': 0.7892904124429312,\n",
       " 'Y': 4.133185629864339,\n",
       " 'R': 4.911479736251563,\n",
       " 'E': 9.608327869119522,\n",
       " 'A': 5.858180233528959,\n",
       " 'L': 10.569079201912542,\n",
       " 'W': 0.9346860147350501,\n",
       " 'I': 7.3015472209890975,\n",
       " 'V': 7.923042148433841,\n",
       " 'Q': 2.04083294982019,\n",
       " 'H': 1.5419264713668408,\n",
       " 'M': 1.865706594118198,\n",
       " 'T': 4.201199819171856,\n",
       " 'N': 3.603530221514476,\n",
       " 'G': 6.7095794116568985}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aquifex_aeolicus_VF5.mean_AA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'K': 2.9440643095706154,\n",
       " 'P': 1.547124235237513,\n",
       " '*': 0.2661738967919316,\n",
       " 'S': 1.7614338898397268,\n",
       " 'D': 1.6051740744098855,\n",
       " 'F': 2.387013257951703,\n",
       " 'C': 1.0981742652812894,\n",
       " 'Y': 1.6028704011973323,\n",
       " 'R': 2.0304425684913956,\n",
       " 'E': 3.283793599655885,\n",
       " 'A': 2.2809981503501815,\n",
       " 'L': 3.080711147423685,\n",
       " 'W': 0.8144887685797604,\n",
       " 'I': 1.99739850196166,\n",
       " 'V': 2.151390016150716,\n",
       " 'Q': 1.2232858395522948,\n",
       " 'H': 1.03805532649652,\n",
       " 'M': 1.0923907381102593,\n",
       " 'T': 1.4265356828926017,\n",
       " 'N': 1.477769120166431,\n",
       " 'G': 2.2531517690362106}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aquifex_aeolicus_VF5.std_AA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'K': 4.516129032258065,\n",
       " 'P': 2.5806451612903225,\n",
       " '*': 0.6451612903225806,\n",
       " 'S': 4.516129032258065,\n",
       " 'D': 4.516129032258065,\n",
       " 'F': 3.225806451612903,\n",
       " 'C': 0.6451612903225806,\n",
       " 'Y': 1.2903225806451613,\n",
       " 'R': 7.096774193548387,\n",
       " 'E': 9.03225806451613,\n",
       " 'A': 12.258064516129032,\n",
       " 'L': 9.67741935483871,\n",
       " 'W': 1.2903225806451613,\n",
       " 'I': 9.67741935483871,\n",
       " 'V': 7.096774193548387,\n",
       " 'Q': 1.2903225806451613,\n",
       " 'H': 2.5806451612903225,\n",
       " 'M': 1.2903225806451613,\n",
       " 'T': 4.516129032258065,\n",
       " 'N': 2.5806451612903225,\n",
       " 'G': 9.67741935483871}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aquifex_aeolicus_VF5['aq_132']['AA_Content_mean']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "eda",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
